lib/llvmopencl/AutomaticLocals.cc

```diff
@@ -213,3 +135,92 @@ AutomaticLocals::processAutomaticLocals(Function *F) {
   return NewKernel;
 }
 
+static bool automaticLocals(Module &M, FunctionVec &OldKernels) {
+  unsigned long ModStrategy = 0;
+  getModuleIntMetadata(M, "device_autolocals_to_args", ModStrategy);
+
+  if (ModStrategy == POCL_AUTOLOCALS_TO_ARGS_NEVER) {
+    return false;
+  }
+  bool changed = false;
+
+  // store the new and old kernel pairs in order to regenerate
+  // all the metadata that used to point to the unmodified
+  // kernels
+  FunctionMapping kernels;
+
+  std::string ErrorInfo;
+  FunctionVec NewFuncs;
+  for (Module::iterator mi = M.begin(), me = M.end(); mi != me; ++mi) {
+    // This is to prevent recursion with llvm 3.9. The new kernels are
+    // recognized as kernelsToProcess.
+    if (find(NewFuncs.begin(), NewFuncs.end(), &*mi) != NewFuncs.end())
+      continue;
+    if (!isKernelToProcess(*mi))
+      continue;
+
+    Function *F = &*mi;
+
+    Function *new_kernel = processAutomaticLocals(F, ModStrategy);
+    if (new_kernel != F)
+      changed = true;
+    kernels[F] = new_kernel;
```

Comment:
Check the other var naming. We can fix the bad C++ style issues now a bit since we touch/move the code around anyhow.